### PR TITLE
[SPARK-11973] [SQL] push filter through aggregation with alias and literals

### PR DESCRIPTION
**Pull request by davies**
Currently, filter can't be pushed through aggregation with alias or literals, this patch fix that.

After this patch, the time of TPC-DS query 4 go down to 13 seconds from 141 seconds (10x improvements).

cc @nongli  @yhuai

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan  too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -690,7 +690,15 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
...
     case filter @ Filter(condition,
         aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition {
+
+      // Create a map of Alias for grouping keys or literals
+      val aliasMap = AttributeMap(aggregateExpressions.collect {
+        case a: Alias if groupingExpressions.contains(a.child) || a.child.foldable =>
+          (a.toAttribute, a.child)
+      })
+      val newCond = PushPredicateThroughProject.replaceAlias(condition, aliasMap)
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think this is confusing to read. Not clear why this is calling a utility in PushPredicateThroughProject. Move this into a better named place (similar to PredicateHelper)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -690,7 +690,15 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
...
+
+      val (pushDown, stayUp) = splitConjunctivePredicates(newCond).partition {
         conjunct => conjunct.references subsetOf AttributeSet(groupingExpressions)
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Can you see if you can make AttributeSet deal with this. It is more general and more consistent with what I understand the intent of AttributeSet is.

The comment in AttributeSet is clear that it wants AttributeReferences which Aliases are not. Is that possible?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -690,7 +690,15 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case filter @ Filter(condition,
         aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition {
+
+      // Create a map of Alias for grouping keys or literals
+      val aliasMap = AttributeMap(aggregateExpressions.collect {
+        case a: Alias if groupingExpressions.contains(a.child) || a.child.foldable =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This doesn't work if a grouping expression is inside another expression, for example, `key + 1 as k`, after remove the alias, `groupingExpressions` doesn't contain `key + 1` and we will fail to push down it.

I think we don't need the `if` here, we have `conjunct.references subsetOf AttributeSet(groupingExpressions)` below to decide whether to push a condition or not.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2028,4 +2028,25 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       Row(false) :: Row(true) :: Nil)
   }
 
+  test("push filter through aggregation with alias and literals") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: should we write test in `FilterPushdownSuite`?

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: +1 

It'd be better to have a unit test for the optimizer ...

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Added

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Do we still need the end-to-end test once we have the unit tests?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -609,6 +609,19 @@ object SimplifyFilters extends Rule[LogicalPlan] {
 }
 
 /**
+  * Helper functions for Predicate push down.
+  */
+object PredicateHelper {
+
+  // Substitute any known alias from a map.
+  def replaceAlias(condition: Expression, aliases: AttributeMap[Expression]): Expression = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: why not move this to the existing `trait PredicateHelper`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -2028,4 +2028,25 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+          | SELECT k, v from (
+          |   SELECT key + 1 AS k, sum(value) + 2 AS v, 3 c FROM src GROUP BY key
+          | ) t WHERE k = 0 and v > 0 and c = 3
+        """.stripMargin)
+      val q2 = sql(
+        """
+          | SELECT k, v from (
+          |   SELECT key + 1 AS k, sum(value) + 2 AS v, 3 c FROM src WHERE key + 1 = 0 GROUP BY key
+          | ) t WHERE v > 0
+        """.stripMargin)
```
Comment (watermen) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Can also push `v > 0` into Subquery use `HAVING`? like:
```
SELECT k, v from (
  SELECT key + 1 AS k, sum(value) + 2 AS v, 3 c FROM src WHERE key + 1 = 0 GROUP BY key HAVING v > 0
) t
```

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: After this patch, the filter in HAVING could also be pushed down, but `v > 0` could not, it depends on AggregateExpression

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -690,12 +684,24 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case filter @ Filter(condition,
         aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition {
-        conjunct => conjunct.references subsetOf AttributeSet(groupingExpressions)
+
+      def hasAggregate(expression: Expression): Boolean = expression match {
+        case agg: AggregateExpression => true
+        case other => expression.children.exists(hasAggregate)
+      }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: a simplification: `expr.find(_.isInstanceOf[AggregateExpression]).isDefined`, but it's minor, may not worth another roundtrip.

**Comment by cloud-fan** [comment]
LGTM

**Comment by davies** [comment]
Merging into master and 1.6 branch.

**Pull request closed without merging**